File: 798-得分最高的最小轮调.cpp
#include <iostream>
using namespace std;
#include <vector>
class Solution
{
public:
    int bestRotation(vector<int> &nums)
    {
        int n = nums.size();
        vector<int> diffs(n);
        for (int i = 0; i < n; i++)
        {
            int low = (i + 1) % n;
            int high = (n - nums[i] + i + 1) % n;
            diffs[low]++;
            diffs[high]--;
            if (low >= high)
            {
                diffs[0]++;
            }
        }
        int max_count = 0, count = 0, idx = 0;
        for (int i = 0; i < n; i++)
        {
            count += diffs[i];
            if (count > max_count)
            {
                idx = i;
                max_count = count;
            }
        }
        return idx;
    }
};

File: 798-得分最高的最小轮调.py
from typing import List
from collections import Counter


class Solution:
    def bestRotation(self, nums: List[int]) -> int:
        n = len(nums)
        # 暴力超时
        # count = {}
        # for i in range(n):
        #     for ss in range(i+1, n-nums[i]+i+1):
        #         ss %= n
        #         if ss not in count:
        #             count[ss] = 1
        #         else:
        #             count[ss] += 1
        # return sorted(count, key=lambda x: (-count[x], x))[0]
        # 差分数组 diff[i] = count[i] - count[i-1]
        diffs = [0] * n
        for i in range(n):
            low = (i+1) % n
            high = (n-nums[i]+i+1) % n
            diffs[low] += 1
            diffs[high] -= 1
            if low >= high:
                diffs[0] += 1
        # 遍历求前缀和得到max_count[i]
        max_count, idx, count = 0, 0, 0
        for i, diff in enumerate(diffs):
            count += diff
            if count > max_count:
                idx, max_count = i, count
        return idx


nums = [2, 3, 1, 4, 0]
sol = Solution().bestRotation(nums)
print(sol)
